chia/server/rate_limits.py

```diff
@@ -129,31 +146,49 @@ def process_msg_and_check(self, message: Message) -> bool:
             log.warning(f"Invalid message: {message.type}, {e}")
             return True
 
-        self.message_counts[message_type] += 1
-        self.message_cumulative_sizes[message_type] += len(message.data)
-        proportion_of_limit = self.percentage_of_limit / 100
-
-        limits = DEFAULT_SETTINGS
-        if message_type in rate_limits_tx:
-            limits = rate_limits_tx[message_type]
-        elif message_type in rate_limits_other:
-            limits = rate_limits_other[message_type]
-            self.non_tx_message_counts += 1
-            self.non_tx_cumulative_size += len(message.data)
-            if self.non_tx_message_counts > NON_TX_FREQ * proportion_of_limit:
+        new_message_counts: int = self.message_counts[message_type] + 1
+        new_cumulative_size: int = self.message_cumulative_sizes[message_type] + len(message.data)
+        new_non_tx_count: int = self.non_tx_message_counts
+        new_non_tx_size: int = self.non_tx_cumulative_size
+        proportion_of_limit: float = self.percentage_of_limit / 100
```

Comment:
this is the key part of this patch. Rather than updating the counters immediately, compute the new counters if this message were to be sent. Check rates against these counters and if we end up accepting sending the message, then commit the new values to the counters (see the `finally` block)

Reply:
This would make a kick ass comment, as it perfectly describes the intent.

Reply:
I was hoping the comment in the `finally` block would suffice